EKF/drag_fusion.cpp

```diff
@@ -82,15 +88,36 @@ void Ekf::fuseDrag()
 
 	// perform sequential fusion of XY specific forces
 	for (uint8_t axis_index = 0; axis_index < 2; axis_index++) {
-		// Estimate the airspeed from the measured drag force and ballistic coefficient
+		// measured drag acceleration corrected for sensor bias
 		const float mea_acc = _drag_sample_delayed.accelXY(axis_index)  - _state.delta_vel_bias(axis_index) / _dt_ekf_avg;
-		const float airSpd = sqrtf((2.0f * fabsf(mea_acc)) / (ballistic_coef_inv_xy(axis_index) * rho));
 
-		// Estimate the derivative of specific force wrt airspeed along the X axis
-		// Limit lower value to prevent arithmetic exceptions
-		const float Kacc = fmaxf(1e-1f, rho * ballistic_coef_inv_xy(axis_index) * airSpd);
+		// predicted drag force sign is opposite to predicted wind relative velocity
+		const float drag_sign = (rel_wind_body(axis_index) >= 0.f) ? 1.f : -1.f;
 
+		float pred_acc; // predicted drag acceleration
 		if (axis_index == 0) {
+			// drag can be modelled as an arbitrary  combination of bluff body drag that proportional to
+			// speed squared, and rotor momentum drag that is proportional to speed.
+			float Kacc; // Derivative of specific force wrt airspeed
+			if (using_mcoef && using_bcoef_x) {
+				// bluff body and propeller momentum drag
+				const float airspeed = (_params.bcoef_x  / rho) * (- _params.mcoef  + sqrtf(sq(_params.mcoef ) + 2.0f * (rho / _params.bcoef_x ) * fabsf(mea_acc)));
+				Kacc = fmaxf(1e-1f, (rho / _params.bcoef_x) * airspeed + _params.mcoef  * density_ratio);
+				pred_acc = (0.5f / _params.bcoef_x ) * rho * sq(rel_wind_body(0)) * drag_sign - rel_wind_body(0) * _params.mcoef  * density_ratio;
```

Comment:
You could still use `ballistic_coef_inv_xy(axis_index)` computed above

Reply:
Done